src/kernel/mpipeline.ml

```diff
@@ -155,19 +235,47 @@ let process
         | (`Interface _ | `Implementation _) as ast ->
           raw_source, Some ast
     )) in
-  let reader = timed_lazy reader_time (lazy (
-      let lazy source = source in
-      let config = Mconfig.normalize config in
-      Mocaml.setup_reader_config config;
-      let result = Mreader.parse ?for_completion config source in
-      result, config
+  let reader =
+    timed_lazy reader_time
+      (lazy
+        (let (lazy ((_, pp_result) as source)) = source in
+          let config = Mconfig.normalize config in
+          Mocaml.setup_reader_config config;
+          let { Reader_with_cache.output; cache_was_hit } =
+            let cache_disabling =
+              match (config.merlin.use_ppx_cache, pp_result) with
+              | false, _ -> Some "configuration"
+              | true, Some _ ->
+                (* The cache could be refined in the future to also act on the PP phase.
+                   For now, let's disable the whole cache when there's a PP. *)
+                Some "source preprocessor usage"
+              | true, None -> None
+            in
+            Reader_with_cache.apply ~cache_disabling ~force_invalidation:false
+              { source; for_completion; config }
+          in
+          { Reader.result = output; config; cache_was_hit }
     )) in
   let ppx = timed_lazy ppx_time (lazy (
-      let lazy ({Mreader.parsetree; _}, config) = reader in
+      let (lazy
+          {
+            Reader.result = { Mreader.parsetree; _ };
+            config;
+            cache_was_hit = source_is_unmodified;
+          }) =
+        reader
+      in
       let caught = ref [] in
       Msupport.catch_errors Mconfig.(config.ocaml.warnings) caught @@ fun () ->
-      let parsetree = Mppx.rewrite config parsetree in
-      { Ppx. config; parsetree; errors = !caught }
+      let {Ppx_with_cache.output = parsetree; cache_was_hit = _ } =
+        let cache_disabling =
+          if config.merlin.use_ppx_cache then None
+          else Some "configuration"
+        in
+        let force_invalidation = not source_is_unmodified in
+        Ppx_with_cache.apply ~cache_disabling ~force_invalidation {parsetree; config}
```

Comment:
Here we also bypass real fingerprinting with the `force_invalidation` parameter.

That means that we get ppx cache misses even for source changes that don't change the parsetree (like adding whitespace).

I don't know if it would be worth it to do it, but we could add a remark about that somewhere.

Reply:
That's room for improvement if the current cache misses too often.

Reply:
Yes, that's a good point. It brings us again to the question of whether it'd be easy to have a digest for the parsetree structure (as opposed to the source code, i.e. string). Do you know whether source code modifications that preserve the parsetree happen more or less frequently? For now, I've just added a comment about this.

Reply:
I think the most frequent cases are whitespace changes: adding or removing lines, adding or removing spaces.
There are also a few syntax that have an equivalent representation.